File: counting_sort.py
def counting_sort(arr):
    maxVal = -1
    for i in arr:
        maxVal = max(maxVal, i)
    freq, accSum = [0]*(maxVal+1), []
    for i in arr:
        freq[i] += 1
    prev = 0
    for j in freq:
        prev += j
        accSum.append(prev)
    res = [0]*len(arr)
    for i in arr:
        idx = accSum[i]-freq[i]
        freq[i] -= 1
        res[idx] = i
    return res


data = [4, 3, 12, 1, 5, 5, 3, 9]

sorted_data = counting_sort(data)
print(sorted_data)


# Time Complexity: O(n+max) where max is largest val in arr
# Space Complexity: O(max)

File: CountingSort.java
import java.util.Arrays;

public class CountingSort {

    public static int[] countingSort(int arr[]) {
        int maxVal = -1;
        for (int i : arr)
            maxVal = Math.max(i, maxVal);
        int freq[] = new int[maxVal + 1];
        for (int i : arr)
            freq[i]++;
        int cumSum[] = new int[maxVal + 1];
        for (int i = 1; i <= maxVal; i++) {
            cumSum[i] += cumSum[i - 1] + freq[i];
        }
        int res[] = new int[arr.length];
        for (int i : arr) {
            int idx = cumSum[i] - freq[i];
            freq[i]--;
            res[idx] = i;
        }
        return res;
    }

    public static void main(String args[]) {

        int[] data = { 4, 3, 12, 1, 5, 5, 3, 9 };
        System.out.println("Sorted Array in Ascending Order:");
        System.out.println(Arrays.toString(countingSort(data)));
    }
}
